cmd/faFindFast/efficient.go

```diff
@@ -4,133 +4,210 @@ import (
-		// enough to have a full "windowSize" of bases in the window
-		if lastRefIdxOfWindow-refIdxBeforeWindow > windowSize {
-			alnIdxBeforeWindow, _, _, _, numRefNs, _, numQueryNsMatch, numSubst = incrementWindowEdge(reference, query, alnIdxBeforeWindow)
-			refIdxBeforeWindow++
+		// usually increment the baseBeforeWindow,
+		// but not at the beginning when we have not yet incremented the end enough to have a full "windowSize" of bases in the window
+		if lastFirstQueryIdxOfWindow-firstQueryIdxBeforeWindow > s.WindowSize {
+			alnIdxBeforeWindow, _, _, _, numFirstQueryNs, _, numSecondQueryNsMatch, numSubst = incrementWindowEdge(firstQuery, secondQuery, alnIdxBeforeWindow)
+			alnIdxBeforeWindowForRef = updateAlnIdxBeforeWindow(firstQuery, alnIdxBeforeWindow)
+			firstQueryIdxBeforeWindow++
 			//totalGaps -= gapOpenCloseRef + gapClosedQuery
-			totalNs -= numRefNs + numQueryNsMatch
+			totalNs -= numFirstQueryNs + numSecondQueryNsMatch
 			totalSubst -= numSubst
 		}
 
-		// the trailing window needs to "look ahead" to see what happens before the next ref base
-		if lastRefIdxOfWindow-refIdxBeforeWindow == windowSize {
-			_, gapOpenCloseRef, _, gapClosedQuery, _, numQueryNsGap, _, _ = incrementWindowEdge(reference, query, alnIdxBeforeWindow)
-			totalGaps -= gapOpenCloseRef + gapClosedQuery
-			totalNs -= numQueryNsGap
+		// the trailing window needs to "look ahead" to see what happens before the next firstQuery base
+		if lastFirstQueryIdxOfWindow-firstQueryIdxBeforeWindow == s.WindowSize {
+			_, gapOpenCloseFirstQuery, _, gapClosedSecondQuery, _, numSecondQueryNsGap, _, _ = incrementWindowEdge(firstQuery, secondQuery, alnIdxBeforeWindow)
+			totalGaps -= gapOpenCloseFirstQuery + gapClosedSecondQuery
+			totalNs -= numSecondQueryNsGap
 		}
 
 		// we check to make sure we are not at the very beginning or end, where we would have partial or illegal windows
-		if lastRefIdxOfWindow-refIdxBeforeWindow == windowSize && lastAlnIdxOfWindow < len(reference) {
-			// an option/flag can tell us not to print if there are Ns in the query or target
-			if !noPrintIfN || totalNs == 0 {
-				if longOutput {
-					percentDiverged = 100 * (float64(totalSubst+totalGaps) / float64(windowSize))
-					if totalSubst+totalGaps > windowSize {
-						log.Fatalf("Error: total number of mutations exceeds windowSize. This may or may not be a bug, but your sequence has deviated from our use case.")
-					}
-					rawPValue = scorePValueCache[totalSubst+totalGaps]
-					_, err = fmt.Fprintf(out, "%s\t%d\t%d\t%s_%d\t%d\t%s\t%e\t%e\n", refChrName, refIdxBeforeWindow+1, lastRefIdxOfWindow+1, refChrName, refIdxBeforeWindow+1, totalSubst+totalGaps, "+", percentDiverged, rawPValue)
-					exception.PanicOnErr(err)
+		if lastFirstQueryIdxOfWindow-firstQueryIdxBeforeWindow == s.WindowSize && lastAlnIdxOfWindow < len(firstQuery) {
+
+			// convert position from firstQuery to reference
```

Comment:
second main new function in this PR: convert position from firstQuery to reference